engine/src/witness/common/epoch_source.rs

```diff
@@ -420,17 +444,39 @@ impl<
 	{
 		self.filter_map(
 			|state_chain_client, epoch, block_hash, info| async move {
-				state_chain_client
+				match state_chain_client
 					.storage_map_entry::<pallet_cf_vaults::VaultStartBlockNumbers<
 						state_chain_runtime::Runtime,
 						<TChain as PalletInstanceAlias>::Instance,
 					>>(block_hash, &epoch)
 					.await
 					.expect(STATE_CHAIN_CONNECTION)
-					.map(|vault| (vault, info))
+				{
+					Some(vault_start_block_number) => {
+						Some((
+						state_chain_client
+						.storage_map_entry::<pallet_cf_threshold_signature::VaultKeys<
+							state_chain_runtime::Runtime,
+							<TChain as PalletInstanceAlias>::Instance,
+						>>(block_hash, &epoch)
+						.await
+						.expect(STATE_CHAIN_CONNECTION)
+						.expect("since the block start number for this epoch exists, the vaulkt key for this epoch should also exist"),
```

Comment:
Previously, it was difficult to show that here since the key and the start number were set separately and you would have to follow the code to convince yourself that it is true, but since the broadcast barrier functionality and optimistic rotation, both these values are set together in one function (```activate_vaults()``````) and so it is easier to convince themselves that it is true. I'll add a reference to the function.